src/data/mv_polynomial.lean

```diff
@@ -1298,4 +1329,115 @@ end
+variable {S : Type}
+
+/-- `pderivative v p` is the partial derivative of `p` with respect to `v` -/
+def pderivative (v : S) (p : mv_polynomial S R) : mv_polynomial S R :=
+p.sum (λ A B, monomial (A - single v 1) (B * (A v)))
+
+
+lemma pderivative_add {v : S} {f g : mv_polynomial S R} :
+  pderivative v (f + g) = pderivative v f + pderivative v g :=
+begin
+  refine sum_add_index _ _,
+  { intro, simpa [monomial] },
+  intros, rw [monomial_add, add_mul],
+end
+
+@[simp]
+lemma pderivative_monomial {v : S} {u : S →₀ ℕ} {a : R} :
+  pderivative v (monomial u a) = monomial (u - single v 1) (a * (u v)) :=
+by simp [pderivative]
+
+lemma pderivative_C {v : S} {a : R} : pderivative v (C a) = 0 :=
+suffices pderivative v (monomial 0 a) = 0, by rw C; exact this,
+by rw [pderivative_monomial, zero_apply, nat.cast_zero, mul_zero, monomial_zero]
+
+lemma pderivative_zero {v : S} : pderivative v (0 : mv_polynomial S R) = 0 :=
+suffices pderivative v (C 0 : mv_polynomial S R) = 0, by rw C_0 at this; exact this,
+show pderivative v (C 0 : mv_polynomial S R) = 0, from pderivative_C
+
+lemma pderivative_monomial_single {a : R} {v : S} {n : ℕ} :
+  pderivative v (monomial (single v n) a) = monomial (single v (n-1)) (a * n) :=
+by simp
+
+private lemma pderivative_monomial_mul_aux1_aux {v : S} {u u' : S →₀ ℕ} (h : u v ≠ 0) :
+  u - single v 1 + u' = u + u' - single v 1 :=
+begin
+  ext f,
+  rw [add_apply, nat_sub_apply, nat_sub_apply, add_apply],
+  by_cases h : v = f,
+  { rw [←h, single_eq_same], cases (u v), { contradiction },
+    rw [nat.succ_sub_one, nat.succ_add, nat.succ_sub_one] },
+  rw single_eq_of_ne h, simp
+end
+
+@[simp]
+private lemma pderivative_monomial_mul_aux1 {v : S} {u u' : S →₀ ℕ} {r r' : R} :
+  monomial (u - single v 1 + u') (r * (u v) * r') =
+    monomial (u + u' - single v 1) (r * (u v) * r') :=
+begin
+  by_cases h : u v = 0,
+  { rw [h, nat.cast_zero, mul_zero, zero_mul], simp [monomial] },
+  rw pderivative_monomial_mul_aux1_aux h,
+end
+
+private lemma pderivative_monomial_mul_aux2_aux {v : S} {u u' : S →₀ ℕ} (h : u' v ≠ 0) :
+  u + (u' - single v 1) = u + u' - single v 1 :=
+begin
+  ext f,
+  rw [add_apply, nat_sub_apply, nat_sub_apply, add_apply],
+  by_cases h : v = f,
+  { rw [←h, single_eq_same], cases (u' v), { contradiction },
+    rw [nat.succ_sub_one, nat.add_succ, nat.succ_sub_one] },
+  rw single_eq_of_ne h, simp,
+end
+
+@[simp]
+private lemma pderivative_monomial_mul_aux2 {v : S} {u u' : S →₀ ℕ} {r r' : R} :
+  monomial (u + (u' - single v 1)) (r * (r' * (u' v))) =
+    monomial (u + u' - single v 1) (r * (r' * (u' v))) :=
+begin
+  by_cases h : u' v = 0,
+  { rw [h, nat.cast_zero, mul_zero, mul_zero], simp [monomial] },
+  rw pderivative_monomial_mul_aux2_aux h,
+end
+
+lemma pderivative_monomial_mul {v : S} {u u' : S →₀ ℕ} {r r' : R} :
+  pderivative v (monomial u r * monomial u' r') =
+    pderivative v (monomial u r) * monomial u' r' + monomial u r * pderivative v (monomial u' r') :=
```

Comment:
We don't have the design quite right yet, as this feels like it ought to be a `simp` lemma, but if you try this and then write `#lint` you'll see the linter rightly complains that it already knows how to simplify the LHS.

We can return to this later. In the end, we may want to reverse the direction of `monomial_add` and `monomial_mul` back to the way they were, and just use the reverse directions in the proofs here.